python/lsst/daf/butler/core/deferredDatasetHandle.py

```diff
@@ -0,0 +1,113 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Module containing classes used with deferring dataset loading
+"""
+
+__all__ = ("DeferredDatasetHandle",)
+
+import dataclasses
+import typing
+import types
+
+from .datasets import DatasetRef, DatasetType
+from .dimensions import DataId
+
+if typing.TYPE_CHECKING:
+    from ..butler import Butler
+
+
+@dataclasses.dataclass(frozen=True)
+class DeferredDatasetHandle:
+    """Class to support deferred loading of a dataset from a butler
+
+    Parameters
+    ----------
+    butler : `Butler`
+        The butler that will be used to fetch the deferred dataset
+    datasetRefOrType : `DatasetRef`, `DatasetType`, or `str`
+        When `DatasetRef` the `dataId` should be `None`.
+        Otherwise the `DatasetType` or name thereof.
+    dataId : `dict` or `DataId`
+        A `dict` of `Dimension` link name, value pairs that label the
+        `DatasetRef` within a Collection. When `None`, a `DatasetRef`
+        should be provided as the first argument.
+    parameters : `dict`
+        Additional StorageClass-defined options to control reading,
+        typically used to efficiently read only a subset of the dataset.
+    kwds : `dict`
+        Additional keyword arguments used to augment or construct a
+        `DataId`.  See `DataId` parameters.
```

Comment:
``dataId`` parameter?

Reply:
No in this case I do mean a DataId object itself